In [47]:
from langchain_community.document_loaders import DirectoryLoader, Docx2txtLoader

# 디렉토리 로더로 모든 docx 파일 자동 로드
directory_loader = DirectoryLoader(
    path="data/",
    glob="*.docx",
    loader_cls=Docx2txtLoader,
    show_progress=True,
    use_multithreading=True
)

# 모든 문서 로드
raw_documents = directory_loader.load()


print(raw_documents)

# # 텍스트 분할
text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=700,
    chunk_overlap=110,
)

all_documents = text_splitter.split_documents(raw_documents)
print(f"📊 총 {len(all_documents)}개 문서 청크 생성!")

100%|██████████| 8/8 [00:03<00:00,  2.64it/s]

[Document(metadata={'source': 'data\\민영교도소 등의 설치ㆍ운영에 관한 법률(법률)(제17504호)(20210421).docx'}, page_content='민영교도소 등의 설치ㆍ운영에 관한 법률\n\n민영교도소 등의 설치ㆍ운영에 관한 법률 ( 약칭: 민영교도소법 )\n\n[시행 2021. 4. 21.] [법률 제17504호, 2020. 10. 20., 일부개정]\n\n법무부(교정기획과) 02-2110-3364~5\n\n\n\n제2조(정의) 이 법에서 사용하는 용어의 뜻은 다음과 같다.\n\n1. “교정업무(矯正業務)”란 「형의 집행 및 수용자의 처우에 관한 법률」 제2조제4호에 따른 수용자(이하 “수용자”라 한다)의 수용ㆍ관리, 교정(矯正)ㆍ교화(敎化), 직업교육, 교도작업(矯導作業), 분류ㆍ처우, 그 밖에 「형의 집행 및 수용자의 처우에 관한 법률」에서 정하는 업무를 말한다.\n\n2. “수탁자(受託者)”란 제3조에 따라 교정업무를 위탁받기로 선정된 자를 말한다.\n\n3. “교정법인”이란 법무부장관으로부터 교정업무를 포괄적으로 위탁받아 교도소ㆍ소년교도소 또는 구치소 및 그 지소(이하 “교도소등”이라 한다)를 설치ㆍ운영하는 법인을 말한다.\n\n4. “민영교도소등”이란 교정법인이 운영하는 교도소등을 말한다.\n\n[전문개정 2009. 3. 25.]\n\n\n\n법제처\n\n\n\n \n\n국가법령정보센터'), Document(metadata={'source': 'data\\성폭력범죄자의 성충동 약물치료에 관한 법률(법률)(제10371호)(20110724).docx'}, page_content='성폭력범죄자의 성충동 약물치료에 관한 법률\n\n성폭력범죄자의 성충동 약물치료에 관한 법률 ( 약칭: 성충동약물치료법 )\n\n[시행 2011. 7. 24.] [법률 제10371호, 2010. 7. 23., 제정]\n\n법무부(치료처우과) 02-2110-3079\n\n\n\n제1장 총칙\n\n\n\n제1조(목적) 이 법은 16세 미만의 사

In [48]:
all_documents

[Document(metadata={'source': 'data\\민영교도소 등의 설치ㆍ운영에 관한 법률(법률)(제17504호)(20210421).docx'}, page_content='민영교도소 등의 설치ㆍ운영에 관한 법률\n\n민영교도소 등의 설치ㆍ운영에 관한 법률 ( 약칭: 민영교도소법 )\n\n[시행 2021. 4. 21.] [법률 제17504호, 2020. 10. 20., 일부개정]\n\n법무부(교정기획과) 02-2110-3364~5\n\n\n\n제2조(정의) 이 법에서 사용하는 용어의 뜻은 다음과 같다.\n\n1. “교정업무(矯正業務)”란 「형의 집행 및 수용자의 처우에 관한 법률」 제2조제4호에 따른 수용자(이하 “수용자”라 한다)의 수용ㆍ관리, 교정(矯正)ㆍ교화(敎化), 직업교육, 교도작업(矯導作業), 분류ㆍ처우, 그 밖에 「형의 집행 및 수용자의 처우에 관한 법률」에서 정하는 업무를 말한다.\n\n2. “수탁자(受託者)”란 제3조에 따라 교정업무를 위탁받기로 선정된 자를 말한다.\n\n3. “교정법인”이란 법무부장관으로부터 교정업무를 포괄적으로 위탁받아 교도소ㆍ소년교도소 또는 구치소 및 그 지소(이하 “교도소등”이라 한다)를 설치ㆍ운영하는 법인을 말한다.\n\n4. “민영교도소등”이란 교정법인이 운영하는 교도소등을 말한다.\n\n[전문개정 2009. 3. 25.]\n\n\n\n법제처\n\n\n\n \n\n국가법령정보센터'),
 Document(metadata={'source': 'data\\성폭력범죄자의 성충동 약물치료에 관한 법률(법률)(제10371호)(20110724).docx'}, page_content='성폭력범죄자의 성충동 약물치료에 관한 법률\n\n성폭력범죄자의 성충동 약물치료에 관한 법률 ( 약칭: 성충동약물치료법 )\n\n[시행 2011. 7. 24.] [법률 제10371호, 2010. 7. 23., 제정]\n\n법무부(치료처우과) 02-2110-3079\n\n\n\n제1장 총칙\n\n\n\n제1조(목적) 이 법은 16세 미만의 

In [58]:
from langchain_chroma import Chroma
from langchain_openai import OpenAIEmbeddings

embeddings = OpenAIEmbeddings(model="text-embedding-3-small")

vector_store = Chroma(
    collection_name="criminal_law_collection",
    persist_directory="./criminal_law_collection",
    embedding_function=embeddings,
)

In [50]:
# 필요한 라이브러리 설치

from datasets import load_dataset
from dotenv import load_dotenv
from huggingface_hub import login
from os import getenv

import json

HF_TOKEN = getenv("HUGGINGFACE_API_KEY")

login(HF_TOKEN)
load_dotenv()

True

In [51]:
# KMMLU 데이터셋 로드
dataset = load_dataset("HAERAE-HUB/KMMLU", "Criminal-Law")
print(dataset)

DatasetDict({
    train: Dataset({
        features: ['question', 'answer', 'A', 'B', 'C', 'D', 'Category', 'Human Accuracy'],
        num_rows: 85
    })
    dev: Dataset({
        features: ['question', 'answer', 'A', 'B', 'C', 'D', 'Category', 'Human Accuracy'],
        num_rows: 5
    })
    test: Dataset({
        features: ['question', 'answer', 'A', 'B', 'C', 'D', 'Category', 'Human Accuracy'],
        num_rows: 200
    })
})


In [52]:
# 테스트 데이터 확인
test_data = dataset["test"]
print(f"📋 테스트 데이터 개수: {len(test_data)}개")

# 첫 번째 문제 확인
first_question = test_data[0]
print(f"\n🔍 첫 번째 문제:")
print(json.dumps(first_question, indent=2, ensure_ascii=False))

📋 테스트 데이터 개수: 200개

🔍 첫 번째 문제:
{
  "question": "범죄 측정에 대한 설명으로 옳은 것은?",
  "answer": 1,
  "A": "참여 관찰 연구는 조사자의 주관적 편견이 개입할 수 있고, 시간과 비용이 많이 들며 연구결과의 일반화가 어렵다 .",
  "B": "인구 대비 범죄 발생 건수를 의미하는 범죄율 (crime rate)은 각 범죄의 가치를 서로 다르게 평가한 다 .",
  "C": "자기 보고식 조사(self-report survey) 는 경미한 범죄보다는 살인 등 중대한 범죄를 측정하는 데 사용된 다 .",
  "D": "피해 조사(victimiz ation survey) 는 개인적 보고에 기반하는 점에서 조사의 객관성과 정확성을 확보할 수 있다.",
  "Category": "Criminal Law",
  "Human Accuracy": 0.0
}


In [81]:
first_question

{'question': '범죄 측정에 대한 설명으로 옳은 것은?',
 'answer': 1,
 'A': '참여 관찰 연구는 조사자의 주관적 편견이 개입할 수 있고, 시간과 비용이 많이 들며 연구결과의 일반화가 어렵다 .',
 'B': '인구 대비 범죄 발생 건수를 의미하는 범죄율 (crime rate)은 각 범죄의 가치를 서로 다르게 평가한 다 .',
 'C': '자기 보고식 조사(self-report survey) 는 경미한 범죄보다는 살인 등 중대한 범죄를 측정하는 데 사용된 다 .',
 'D': '피해 조사(victimiz ation survey) 는 개인적 보고에 기반하는 점에서 조사의 객관성과 정확성을 확보할 수 있다.',
 'Category': 'Criminal Law',
 'Human Accuracy': 0.0}

In [93]:
test_samples = []
for i in range(len(test_data)):
    sample = {
        'question': test_data[i]['question'],
        'A': test_data[i]['A'],
        'B': test_data[i]['B'], 
        'C': test_data[i]['C'],
        'D': test_data[i]['D'],
    }
    test_samples.append(sample)

# 'answer': test_data[i]['answer'],
# 'category': test_data[i]['Category'],
# 'human_accuracy': test_data[i]['Human Accuracy']

print(f"변환 완료: {len(test_samples)}개 문제")
print(f"첫 번째 문제: {test_samples[0]}")

question = test_samples[:5]

변환 완료: 200개 문제
첫 번째 문제: {'question': '범죄 측정에 대한 설명으로 옳은 것은?', 'A': '참여 관찰 연구는 조사자의 주관적 편견이 개입할 수 있고, 시간과 비용이 많이 들며 연구결과의 일반화가 어렵다 .', 'B': '인구 대비 범죄 발생 건수를 의미하는 범죄율 (crime rate)은 각 범죄의 가치를 서로 다르게 평가한 다 .', 'C': '자기 보고식 조사(self-report survey) 는 경미한 범죄보다는 살인 등 중대한 범죄를 측정하는 데 사용된 다 .', 'D': '피해 조사(victimiz ation survey) 는 개인적 보고에 기반하는 점에서 조사의 객관성과 정확성을 확보할 수 있다.'}


In [98]:
question

[{'question': '범죄 측정에 대한 설명으로 옳은 것은?',
  'A': '참여 관찰 연구는 조사자의 주관적 편견이 개입할 수 있고, 시간과 비용이 많이 들며 연구결과의 일반화가 어렵다 .',
  'B': '인구 대비 범죄 발생 건수를 의미하는 범죄율 (crime rate)은 각 범죄의 가치를 서로 다르게 평가한 다 .',
  'C': '자기 보고식 조사(self-report survey) 는 경미한 범죄보다는 살인 등 중대한 범죄를 측정하는 데 사용된 다 .',
  'D': '피해 조사(victimiz ation survey) 는 개인적 보고에 기반하는 점에서 조사의 객관성과 정확성을 확보할 수 있다.'},
 {'question': '국가의 기능과 관련한 죄에 대한 설명으로 옳은 것은?',
  'A': '甲이 자기 자신을 무고하기로 제3자와 공모하고 이에 따라 무고 행위에 가담한 경우, 甲에게 무고죄의 공동정범이 성립한다.',
  'B': '甲이 허위로 신고한 사실이 무고행위 당시에는 형사처분의 대상이 될 수 있었으나, 이후 그러한 사실이 형사범죄가 되지 않는 것으로 판례가 변경된 경우, 甲에게 무고죄가 성립하지 않는다.',
  'C': '甲이 자신에 대한 형사처분이나 징계 처분을 피하기 위하여 증거를 인멸한 것이 동시에 다른 공범자의 증거를 인멸한 결과가 된 경우, 甲에게 증거인멸죄가 성립한다.',
  'D': '공무원인 甲이 직무 관련자에게 제3자와 계약을 체결하도록 요구하여 계약 체결을 하게 한 행위가 제3자 뇌물수수죄의 구성요건과 직권남용권리행사방해죄의 구성요건에 모두 해당하는 경우, 제3자 뇌물수수죄와 직권남용권리행사방해죄는 상상적 경합의 관계에 있다.'},
 {'question': '甲은 처乙과 부부싸움을 하다가 화가 나서 폭행의 고의로 乙의 가슴을 세게 밀쳤고, 乙은 그 충격으로 사망하고 말았다 이후 甲은 자신의 집이 화재보험에 들어있다는 사실을 인지하고, 범행을 은폐하기 위하여 탁자에 불을 붙인 후 밖으로 나와버렸고 집은 전소되었다 그

In [99]:
for question, A in enumerate(question):
    print(question, A)


0 {'question': '범죄 측정에 대한 설명으로 옳은 것은?', 'A': '참여 관찰 연구는 조사자의 주관적 편견이 개입할 수 있고, 시간과 비용이 많이 들며 연구결과의 일반화가 어렵다 .', 'B': '인구 대비 범죄 발생 건수를 의미하는 범죄율 (crime rate)은 각 범죄의 가치를 서로 다르게 평가한 다 .', 'C': '자기 보고식 조사(self-report survey) 는 경미한 범죄보다는 살인 등 중대한 범죄를 측정하는 데 사용된 다 .', 'D': '피해 조사(victimiz ation survey) 는 개인적 보고에 기반하는 점에서 조사의 객관성과 정확성을 확보할 수 있다.'}
1 {'question': '국가의 기능과 관련한 죄에 대한 설명으로 옳은 것은?', 'A': '甲이 자기 자신을 무고하기로 제3자와 공모하고 이에 따라 무고 행위에 가담한 경우, 甲에게 무고죄의 공동정범이 성립한다.', 'B': '甲이 허위로 신고한 사실이 무고행위 당시에는 형사처분의 대상이 될 수 있었으나, 이후 그러한 사실이 형사범죄가 되지 않는 것으로 판례가 변경된 경우, 甲에게 무고죄가 성립하지 않는다.', 'C': '甲이 자신에 대한 형사처분이나 징계 처분을 피하기 위하여 증거를 인멸한 것이 동시에 다른 공범자의 증거를 인멸한 결과가 된 경우, 甲에게 증거인멸죄가 성립한다.', 'D': '공무원인 甲이 직무 관련자에게 제3자와 계약을 체결하도록 요구하여 계약 체결을 하게 한 행위가 제3자 뇌물수수죄의 구성요건과 직권남용권리행사방해죄의 구성요건에 모두 해당하는 경우, 제3자 뇌물수수죄와 직권남용권리행사방해죄는 상상적 경합의 관계에 있다.'}
2 {'question': '甲은 처乙과 부부싸움을 하다가 화가 나서 폭행의 고의로 乙의 가슴을 세게 밀쳤고, 乙은 그 충격으로 사망하고 말았다 이후 甲은 자신의 집이 화재보험에 들어있다는 사실을 인지하고, 범행을 은폐하기 위하여 탁자에 불을 붙인 후 밖으로 나와버렸고 집은 전소되었다 그 후 甲은 경찰에 자신이 집

In [62]:
retriever = vector_store.as_retriever(search_kwargs={"k": 4})

result = retriever.get_relevant_documents(first_question["question"])

In [92]:
result

[Document(id='9c923926-a80f-48cb-bf98-14cf8053f2c1', metadata={'source': 'data\\특정 범죄자에 대한 위치추적 전자장치 부착 등에 관한 법률(법률)(제10257호)(20100716).docx'}, page_content='특정 범죄자에 대한 위치추적 전자장치 부착 등에 관한 법률\n\n특정 범죄자에 대한 위치추적 전자장치 부착 등에 관한 법률 ( 약칭: 전자장치부착법 )\n\n[시행 2010. 7. 16.] [법률 제10257호, 2010. 4. 15., 일부개정]\n\n법무부(치료처우과) 02-2110-3330\n\n\n\n제1장 총칙\n\n\n\n제1조(목적) 이 법은 특정 범죄자의 재범 방지와 성행(性行)교정을 통한 재사회화를 위하여 그의 행적을 추적하여 위치를 확인할 수 있는 전자장치를 신체에 부착하게 하는 부가적인 조치를 취함으로써 특정범죄로부터 국민을 보호함을 목적으로 한다. <개정 2009. 5. 8.>\n\n\n\n제2조(정의) 이 법에서 사용하는 용어의 정의는 다음과 같다. <개정 2007. 8. 3., 2009. 5. 8., 2009. 6. 9., 2010. 4. 15.>\n\n1. “특정범죄”란 성폭력범죄, 미성년자 대상 유괴범죄와 살인범죄를 말한다.\n\n2. “성폭력범죄”란 다음 각 목의 범죄를 말한다.'),
 Document(id='07dce6f4-9596-44af-b5a5-a09c7312b624', metadata={'source': 'data\\특정 범죄자에 대한 위치추적 전자장치 부착 등에 관한 법률(법률)(제10257호)(20100716).docx'}, page_content='2. 특정범죄사건에 대하여 무죄(심신상실을 이유로 치료감호가 선고된 경우는 제외한다)ㆍ면소ㆍ공소기각의 판결 또는 결정을 선고하는 때\n\n3. 특정범죄사건에 대하여 벌금형을 선고하는 때\n\n4. 특정범죄사건에 대하여 선고유예 또는 집행유예를 선고하는 때(제28조제1항에 따라 전자장치 부착을 명

In [76]:
source = [i.page_content for i in result]
source

['특정 범죄자에 대한 위치추적 전자장치 부착 등에 관한 법률\n\n특정 범죄자에 대한 위치추적 전자장치 부착 등에 관한 법률 ( 약칭: 전자장치부착법 )\n\n[시행 2010. 7. 16.] [법률 제10257호, 2010. 4. 15., 일부개정]\n\n법무부(치료처우과) 02-2110-3330\n\n\n\n제1장 총칙\n\n\n\n제1조(목적) 이 법은 특정 범죄자의 재범 방지와 성행(性行)교정을 통한 재사회화를 위하여 그의 행적을 추적하여 위치를 확인할 수 있는 전자장치를 신체에 부착하게 하는 부가적인 조치를 취함으로써 특정범죄로부터 국민을 보호함을 목적으로 한다. <개정 2009. 5. 8.>\n\n\n\n제2조(정의) 이 법에서 사용하는 용어의 정의는 다음과 같다. <개정 2007. 8. 3., 2009. 5. 8., 2009. 6. 9., 2010. 4. 15.>\n\n1. “특정범죄”란 성폭력범죄, 미성년자 대상 유괴범죄와 살인범죄를 말한다.\n\n2. “성폭력범죄”란 다음 각 목의 범죄를 말한다.',
 '2. 특정범죄사건에 대하여 무죄(심신상실을 이유로 치료감호가 선고된 경우는 제외한다)ㆍ면소ㆍ공소기각의 판결 또는 결정을 선고하는 때\n\n3. 특정범죄사건에 대하여 벌금형을 선고하는 때\n\n4. 특정범죄사건에 대하여 선고유예 또는 집행유예를 선고하는 때(제28조제1항에 따라 전자장치 부착을 명하는 때를 제외한다)\n\n⑤부착명령 청구사건의 판결은 특정범죄사건의 판결과 동시에 선고하여야 한다.<개정 2009. 5. 8., 2010. 4. 15.>\n\n⑥부착명령 선고의 판결이유에는 요건으로 되는 사실, 증거의 요지 및 적용 법조를 명시하여야 한다.<개정 2010. 4. 15.>\n\n⑦부착명령의 선고는 특정범죄사건의 양형에 유리하게 참작되어서는 아니 된다.<개정 2009. 5. 8., 2010. 4. 15.>\n\n⑧특정범죄사건의 판결에 대하여 상소 및 상소의 포기ㆍ취하가 있는 때에는 부착명령 청구사건의 판결에 대하여도 상소 및 상소의 포기ㆍ취

['특정 범죄자에 대한 위치추적 전자장치 부착 등에 관한 법률\n\n특정 범죄자에 대한 위치추적 전자장치 부착 등에 관한 법률 ( 약칭: 전자장치부착법 )\n\n[시행 2010. 7. 16.] [법률 제10257호, 2010. 4. 15., 일부개정]\n\n법무부(치료처우과) 02-2110-3330\n\n\n\n제1장 총칙\n\n\n\n제1조(목적) 이 법은 특정 범죄자의 재범 방지와 성행(性行)교정을 통한 재사회화를 위하여 그의 행적을 추적하여 위치를 확인할 수 있는 전자장치를 신체에 부착하게 하는 부가적인 조치를 취함으로써 특정범죄로부터 국민을 보호함을 목적으로 한다. <개정 2009. 5. 8.>\n\n\n\n제2조(정의) 이 법에서 사용하는 용어의 정의는 다음과 같다. <개정 2007. 8. 3., 2009. 5. 8., 2009. 6. 9., 2010. 4. 15.>\n\n1. “특정범죄”란 성폭력범죄, 미성년자 대상 유괴범죄와 살인범죄를 말한다.\n\n2. “성폭력범죄”란 다음 각 목의 범죄를 말한다.',
 '2. 특정범죄사건에 대하여 무죄(심신상실을 이유로 치료감호가 선고된 경우는 제외한다)ㆍ면소ㆍ공소기각의 판결 또는 결정을 선고하는 때\n\n3. 특정범죄사건에 대하여 벌금형을 선고하는 때\n\n4. 특정범죄사건에 대하여 선고유예 또는 집행유예를 선고하는 때(제28조제1항에 따라 전자장치 부착을 명하는 때를 제외한다)\n\n⑤부착명령 청구사건의 판결은 특정범죄사건의 판결과 동시에 선고하여야 한다.<개정 2009. 5. 8., 2010. 4. 15.>\n\n⑥부착명령 선고의 판결이유에는 요건으로 되는 사실, 증거의 요지 및 적용 법조를 명시하여야 한다.<개정 2010. 4. 15.>\n\n⑦부착명령의 선고는 특정범죄사건의 양형에 유리하게 참작되어서는 아니 된다.<개정 2009. 5. 8., 2010. 4. 15.>\n\n⑧특정범죄사건의 판결에 대하여 상소 및 상소의 포기ㆍ취하가 있는 때에는 부착명령 청구사건의 판결에 대하여도 상소 및 상소의 포기ㆍ취하가 있는 것으로 본다. 상소권회복 또는 재심의 청구나 비상상고가 있는 때에도 또한 같다.<개정 2009. 5. 8., 2010. 4. 15.>\n\n⑨제8항에도 불구하고 검사 또는 피부착명령청구자 및 「형사소송법」 제340조ㆍ제341조에 규정된 자는 부착명령에 대하여 독립하여 상소 및 상소의 포기ㆍ취하를 할 수 있다. 상소권회복 또는 재심의 청구나 비상상고의 경우에도 또한 같다.<개정 2010. 4. 15.>',
 '3. 성폭력범죄를 2회 이상 범하여(유죄의 확정판결을 받은 경우를 포함한다) 그 습벽이 인정된 때\n\n4. 16세 미만의 사람에 대하여 성폭력범죄를 저지른 때\n\n② 검사는 미성년자 대상 유괴범죄를 저지른 사람으로서 미성년자 대상 유괴범죄를 다시 범할 위험성이 있다고 인정되는 사람에 대하여 부착명령을 법원에 청구할 수 있다. 다만, 유괴범죄로 징역형의 실형 이상의 형을 선고받아 그 집행이 종료 또는 면제된 후 다시 유괴범죄를 저지른 경우에는 부착명령을 청구하여야 한다.<신설 2009. 5. 8., 2010. 4. 15.>\n\n③ 검사는 살인범죄를 저지른 사람으로서 살인범죄를 다시 범할 위험성이 있다고 인정되는 사람에 대하여 부착명령을 법원에 청구할 수 있다. 다만, 살인범죄로 징역형의 실형 이상의 형을 선고받아 그 집행이 종료 또는 면제된 후 다시 살인범죄를 저지른 경우에는 부착명령을 청구하여야 한다.<신설 2010. 4. 15.>\n\n④제1항부터 제3항까지의 규정에 따른 부착명령의 청구는 공소가 제기된 특정범죄사건의 항소심 변론종결 시까지 하여야 한다.<개정 2009. 5. 8., 2010. 4. 15.>\n\n⑤법원은 공소가 제기된 특정범죄사건을 심리한 결과 부착명령을 선고할 필요가 있다고 인정하는 때에는 검사에게 부착명령의 청구를 요구할 수 있다.<개정 2009. 5. 8., 2010. 4. 15.>',
 '1. “특정범죄”란 성폭력범죄, 미성년자 대상 유괴범죄와 살인범죄를 말한다.\n\n2. “성폭력범죄”란 다음 각 목의 범죄를 말한다.\n\n가. 「형법」 제2편제32장 강간과 추행의 죄 중 제297조(강간)ㆍ제298조(강제추행)ㆍ제299조(준강간, 준강제추행)ㆍ제300조(미수범)ㆍ제301조(강간등 상해ㆍ치상)ㆍ제301조의2(강간등 살인ㆍ치사)ㆍ제302조(미성년자등에 대한 간음)ㆍ제303조(업무상위력등에 의한 간음)ㆍ제305조(미성년자에 대한 간음, 추행), 제2편제38장 절도와 강도의 죄 중 제339조(강도강간) 및 제340조(해상강도)제3항(부녀를 강간한 죄만을 말한다)의 죄\n\n나. 「성폭력범죄의 처벌 등에 관한 특례법」 제3조(특수강도강간 등)부터 제10조(업무상 위력 등에 의한 추행)까지의 죄 및 제14조(미수범)의 죄(제3조부터 제9조까지의 미수범만을 말한다)\n\n다. 「아동ㆍ청소년의 성보호에 관한 법률」 제7조(아동ㆍ청소년에 대한 강간, 강제추행등)의 죄\n\n라. 가목부터 다목까지의 죄로서 다른 법률에 따라 가중 처벌되는 죄\n\n3. “미성년자 대상 유괴범죄”란 다음 각 목의 범죄를 말한다.']

In [86]:
from langchain import hub

rag_prompt = hub.pull("rlm/rag-prompt")
rag_prompt

ChatPromptTemplate(input_variables=['context', 'question'], input_types={}, partial_variables={}, metadata={'lc_hub_owner': 'rlm', 'lc_hub_repo': 'rag-prompt', 'lc_hub_commit_hash': '50442af133e61576e74536c6556cefe1fac147cad032f4377b60c436e6cdcb6e'}, messages=[HumanMessagePromptTemplate(prompt=PromptTemplate(input_variables=['context', 'question'], input_types={}, partial_variables={}, template="You are an assistant for question-answering tasks. Use the following pieces of retrieved context to answer the question. If you don't know the answer, just say that you don't know. Use three sentences maximum and keep the answer concise.\nQuestion: {question} \nContext: {context} \nAnswer:"), additional_kwargs={})])

template="You are an assistant for question-answering tasks. Use the following pieces of retrieved context to answer the question. If you don't know the answer, just say that you don't know. Use three sentences maximum and keep the answer concise.\nQuestion: {question} \nContext: {context} \nAnswer:"

In [88]:
from langchain_openai import ChatOpenAI
from langchain_core.prompts import PromptTemplate
from typing import Literal
from typing_extensions import TypedDict

llm = ChatOpenAI(model="gpt-4o-mini")

class Router(TypedDict):
    answer: Literal["A", "B", "C", "D"]

chain = rag_prompt | llm.with_structured_output(Router)

response = chain.invoke({"question": question, "context": source})
response

{'answer': '정답은 A입니다. 참여 관찰 연구는 조사자의 주관적 편견이 개입할 가능성이 높고, 시간과 비용이 많이 들며 연구 결과의 일반화가 어렵다는 단점이 있습니다. 나머지 선택지들은 모두 잘못된 설명입니다.'}

In [60]:
from langgraph.graph import StateGraph, MessagesState

class State(MessagesState):
    question: str

In [ ]:



def get_prompt(state: State) -> PromptTemplate:

    question = state["question"]
    

    context = retriever.get_relevant_documents(question)

    prompt = PromptTemplate(
        template="""
        법률 문제에 대한 답변을 찾아주세요.
        문제: {question}
        
        """,
    )